File: crypto/block/validator-auth-config.cpp
#include "tos/tos-types.h"
#include "vm/cells/CellSlice.h"

#include "validator-auth-config.h"
#include "validator-auth-profile.h"
namespace block {
namespace {
struct Header {
  bool enabled = false;
  td::Bits256 domain;
  std::uint64_t revision = 0;
  td::Ref<vm::Cell> identities, keys;
};
td::Result<Header> read_registry(td::Ref<vm::Cell> registry) {
  Header header;
  header.enabled = true;
  if (registry.is_null() || registry->get_level())
    return td::Status::Error("validator-auth-registry");
  vm::CellSlice root{vm::NoVm{}, registry};
  if (!root.is_valid() || root.is_special() || root.size() != 880 || root.size_refs() != 4 ||
      root.fetch_ulong(32) != 0x76617131 || root.fetch_ulong(16) != 1)
    return td::Status::Error("validator-auth-registry");
  td::Bits256 fingerprint, policy;
  if (!root.fetch_bits_to(header.domain) || header.domain.is_zero() || !root.fetch_bits_to(fingerprint) ||
      fingerprint.as_slice() !=
          td::Slice(validator_auth_profile_fingerprint.data(), validator_auth_profile_fingerprint.size()))
    return td::Status::Error("validator-auth-profile");
  header.revision = root.fetch_ulong(64);
  if (!root.fetch_bits_to(policy) || policy.is_zero())
    return td::Status::Error("validator-auth-registry");
  header.identities = root.fetch_ref();
  header.keys = root.fetch_ref();
  auto dictionary_shape = [](td::Ref<vm::Cell> cell) {
    vm::CellSlice s{vm::NoVm{}, cell};
    return s.is_valid() && !s.is_special() && s.size() == 1 && s.size_refs() == s.prefetch_ulong(1);
  };
  if (!dictionary_shape(header.identities) || !dictionary_shape(header.keys) || !dictionary_shape(root.fetch_ref()))
    return td::Status::Error("validator-auth-dictionary");
  vm::CellSlice control{vm::NoVm{}, root.fetch_ref()};
  if (!control.is_valid() || control.is_special() || control.size() != 32 || control.size_refs() != 2 ||
      control.fetch_ulong(32) != 0x76616331 || !dictionary_shape(control.fetch_ref()) ||
      !dictionary_shape(control.fetch_ref()))
    return td::Status::Error("validator-auth-control");
  return header;
}
td::Result<Header> read(vm::Dictionary& config) {
  Header header;
  auto capability = config.lookup_ref(td::BitArray<32>{8});
  if (capability.is_null())
    return header;
  vm::CellSlice cap{vm::NoVm{}, capability};
  if (!cap.is_valid() || cap.is_special() || cap.size() != 104 || cap.size_refs() || cap.fetch_ulong(8) != 0xc4)
    return td::Status::Error("validator-auth-config8");
  auto version = cap.fetch_ulong(32), flags = cap.fetch_ulong(64);
  if (!(flags & tos::capValidatorAuth))
    return header;
  header.enabled = true;
  if (version < 16)
    return td::Status::Error("validator-auth-version");
  for (int index : {9, 10}) {
    auto root = config.lookup_ref(td::BitArray<32>{index});
    if (root.is_null())
      return td::Status::Error("validator-auth-required");
    vm::Dictionary required(root, 32);
    auto entry = required.lookup(td::BitArray<32>{46});
    if (entry.is_null() || entry->size() || entry->size_refs())
      return td::Status::Error("validator-auth-required");
  }
  auto counts = config.lookup_ref(td::BitArray<32>{16});
  if (counts.is_null())
    return td::Status::Error("validator-auth-count");
  vm::CellSlice count{vm::NoVm{}, counts};
  if (!count.is_valid() || count.is_special() || count.size() != 48 || count.size_refs())
    return td::Status::Error("validator-auth-count");
  auto maximum = count.fetch_ulong(16), main = count.fetch_ulong(16), minimum = count.fetch_ulong(16);
  if (maximum > 400 || main > maximum || minimum < 1 || minimum > main)
    return td::Status::Error("validator-auth-count");
  return read_registry(config.lookup_ref(td::BitArray<32>{46}));
}
}  // namespace
td::Status validate_validator_auth_root_shape(td::Ref<vm::Cell> root) {
  try {
    auto result = read_registry(std::move(root));
    return result.is_error() ? result.move_as_error() : td::Status::OK();
  } catch (const vm::VmError&) {
    return td::Status::Error("validator-auth-config-cell");
  } catch (const vm::VmVirtError&) {
    return td::Status::Error("validator-auth-config-pruned");
  }
}
td::Status validate_validator_auth_config(vm::Dictionary& config) {
  try {
    auto result = read(config);
    return result.is_error() ? result.move_as_error() : td::Status::OK();
  } catch (const vm::VmError&) {
    return td::Status::Error("validator-auth-config-cell");
  } catch (const vm::VmVirtError&) {
    return td::Status::Error("validator-auth-config-pruned");
  }
}
td::Status validate_validator_auth_transition(vm::Dictionary& before, vm::Dictionary& after) {
  TRY_RESULT(old, read(before));
  TRY_RESULT(next, read(after));
  if (!old.enabled && !next.enabled)
    return td::Status::OK();
  if (!old.enabled)
    return td::Status::Error("validator-auth-transition-unapproved");
  if (!next.enabled)
    return td::Status::Error("validator-auth-downgrade");
  if (old.domain != next.domain)
    return td::Status::Error("validator-auth-domain");
  bool changed =
      old.identities->get_hash() != next.identities->get_hash() || old.keys->get_hash() != next.keys->get_hash();
  if (next.revision < old.revision || next.revision - old.revision != std::uint64_t(changed))
    return td::Status::Error("validator-auth-revision");
  return td::Status::OK();
}
}  // namespace block

File: tosctl/src/block/src/validator_auth_config.rs
use crate::validator_auth_profile::VALIDATOR_AUTH_PROFILE_FINGERPRINT;
use crate::{error, Cell, CellType, ConfigParams, Result, Serializable, SliceData};

struct Header {
    domain: Vec<u8>,
    revision: u64,
    identities: Cell,
    keys: Cell,
}
fn need(ok: bool, reason: &str) -> Result<()> {
    if ok {
        Ok(())
    } else {
        Err(error!("{}", reason))
    }
}
fn param(config: &ConfigParams, at: u32) -> Result<Option<Cell>> {
    match config.config_params.get(at.write_to_bitstring()?)? {
        None => Ok(None),
        Some(value) => {
            need(
                value.remaining_bits() == 0 && value.remaining_references() == 1,
                "validator-auth-parameter",
            )?;
            Ok(Some(value.reference(0)?))
        }
    }
}
fn required(config: &ConfigParams, at: u32, reason: &str) -> Result<Cell> {
    param(config, at)?.ok_or_else(|| error!("{}", reason))
}
fn read_registry(registry: Cell) -> Result<Header> {
    need(registry.level_mask().mask() == 0, "validator-auth-registry")?;
    let mut root = SliceData::load_cell(registry)?;
    need(
        root.cell_type() == CellType::Ordinary
            && root.remaining_bits() == 880
            && root.remaining_references() == 4
            && root.get_next_u32()? == 0x76617131
            && root.get_next_u16()? == 1,
        "validator-auth-registry",
    )?;
    let domain = root.get_next_bits(256)?;
    need(
        domain != [0; 32] && root.get_next_bits(256)? == VALIDATOR_AUTH_PROFILE_FINGERPRINT,
        "validator-auth-profile",
    )?;
    let revision = root.get_next_u64()?;
    need(root.get_next_bits(256)? != [0; 32], "validator-auth-registry")?;
    let identities = root.checked_drain_reference()?;
    let keys = root.checked_drain_reference()?;
    let dictionary_shape = |cell: Cell| -> Result<bool> {
        let mut s = SliceData::load_cell(cell)?;
        Ok(s.cell_type() == CellType::Ordinary
            && s.remaining_bits() == 1
            && s.remaining_references() == usize::from(s.get_next_bit()?))
    };
    need(
        dictionary_shape(identities.clone())?
            && dictionary_shape(keys.clone())?
            && dictionary_shape(root.checked_drain_reference()?)?,
        "validator-auth-dictionary",
    )?;
    let mut control = SliceData::load_cell(root.checked_drain_reference()?)?;
    need(
        control.cell_type() == CellType::Ordinary
            && control.remaining_bits() == 32
            && control.remaining_references() == 2
            && control.get_next_u32()? == 0x76616331
            && dictionary_shape(control.checked_drain_reference()?)?
            && dictionary_shape(control.checked_drain_reference()?)?,
        "validator-auth-control",
    )?;
    Ok(Header { domain, revision, identities, keys })
}
fn read(config: &ConfigParams) -> Result<Option<Header>> {
    let Some(cap) = param(config, 8)? else {
        return Ok(None);
    };
    let mut cap = SliceData::load_cell(cap)?;
    need(
        cap.cell_type() == CellType::Ordinary
            && cap.remaining_bits() == 104
            && cap.remaining_references() == 0
            && cap.get_next_byte()? == 0xc4,
        "validator-auth-config8",
    )?;
    let version = cap.get_next_u32()?;
    if cap.get_next_u64()? & 1024 == 0 {
        return Ok(None);
    }
    need(version >= 16, "validator-auth-version")?;
    for at in [9, 10] {
        let dictionary = crate::HashmapE::with_hashmap(
            32,
            Some(required(config, at, "validator-auth-required")?),
        );
        let value = dictionary
            .get(46u32.write_to_bitstring()?)?
            .ok_or_else(|| error!("validator-auth-required"))?;
        need(
            value.remaining_bits() == 0 && value.remaining_references() == 0,
            "validator-auth-required",
        )?;
    }
    let mut count = SliceData::load_cell(required(config, 16, "validator-auth-count")?)?;
    need(
        count.cell_type() == CellType::Ordinary
            && count.remaining_bits() == 48
            && count.remaining_references() == 0,
        "validator-auth-count",
    )?;
    let maximum = count.get_next_u16()?;
    let main = count.get_next_u16()?;
    let minimum = count.get_next_u16()?;
    need(
        maximum <= 400 && main <= maximum && minimum >= 1 && minimum <= main,
        "validator-auth-count",
    )?;
    read_registry(required(config, 46, "validator-auth-registry")?).map(Some)
}
/// Structural admission accompanies native execution authorization and complete
/// registry validation; it cannot supply either of those independent authorities.
pub fn validate_validator_auth_config(config: &ConfigParams) -> Result<()> {
    read(config).map(|_| ())
}

pub fn validate_validator_auth_transition(
    before: &ConfigParams,
    after: &ConfigParams,
) -> Result<()> {
    let old = read(before)?;
    let next = read(after)?;
    let (old, next) = match (old, next) {
        (None, None) => return Ok(()),
        (None, Some(_)) => return Err(error!("validator-auth-transition-unapproved")),
        (Some(_), None) => return Err(error!("validator-auth-downgrade")),
        (Some(old), Some(next)) => (old, next),
    };
    need(old.domain == next.domain, "validator-auth-domain")?;
    let changed = old.identities.repr_hash() != next.identities.repr_hash()
        || old.keys.repr_hash() != next.keys.repr_hash();
    need(
        next.revision.checked_sub(old.revision) == Some(u64::from(changed)),
        "validator-auth-revision",
    )?;
    Ok(())
}

/// Bounded native root/header admission. Complete archive semantics belong to
/// validated checkpoint loading and authenticated operation replay.
pub fn validate_validator_auth_root_shape(root: &Cell) -> Result<()> {
    read_registry(root.clone()).map(|_| ())
}
